src/plugins/data/common/search/expressions/kibana_context.ts

```diff
@@ -40,75 +48,102 @@ const mergeQueries = (first: Query | Query[] = [], second: Query | Query[]) =>
+  ) => Promise<KibanaContextStartDependencies>
+) => {
+  const kibanaContextFunction: ExpressionFunctionKibanaContext = {
+    name: 'kibana_context',
+    type: 'kibana_context',
+    inputTypes: ['kibana_context', 'null'],
+    help: i18n.translate('data.search.functions.kibana_context.help', {
+      defaultMessage: 'Updates kibana global context',
+    }),
+    args: {
+      q: {
+        types: ['kibana_query', 'null'],
+        aliases: ['query', '_'],
+        default: null,
+        help: i18n.translate('data.search.functions.kibana_context.q.help', {
+          defaultMessage: 'Specify Kibana free form text query',
+        }),
+      },
+      filters: {
+        types: ['kibana_filter', 'null'],
+        multi: true,
+        help: i18n.translate('data.search.functions.kibana_context.filters.help', {
+          defaultMessage: 'Specify Kibana generic filters',
+        }),
+      },
+      timeRange: {
+        types: ['timerange', 'null'],
+        default: null,
+        help: i18n.translate('data.search.functions.kibana_context.timeRange.help', {
+          defaultMessage: 'Specify Kibana time range filter',
+        }),
+      },
+      savedSearchId: {
+        types: ['string', 'null'],
+        default: null,
+        help: i18n.translate('data.search.functions.kibana_context.savedSearchId.help', {
+          defaultMessage: 'Specify saved search ID to be used for queries and filters',
+        }),
+      },
     },
-    timeRange: {
-      types: ['timerange', 'null'],
-      default: null,
-      help: i18n.translate('data.search.functions.kibana_context.timeRange.help', {
-        defaultMessage: 'Specify Kibana time range filter',
-      }),
+
+    extract(state) {
+      const references: SavedObjectReference[] = [];
+      if (state.savedSearchId.length && typeof state.savedeSearchId[0] === 'string') {
+        const refName = 'kibana_context.savedSearchId';
+        references.push({
+          name: refName,
+          type: 'search',
+          id: state.savedSearchId[0] as string,
+        });
+        state.savedSearchId[0] = refName;
+      }
+      return { state, references };
     },
-    savedSearchId: {
-      types: ['string', 'null'],
-      default: null,
-      help: i18n.translate('data.search.functions.kibana_context.savedSearchId.help', {
-        defaultMessage: 'Specify saved search ID to be used for queries and filters',
-      }),
+
+    inject(state, references) {
+      const reference = references.find((r) => r.name === 'kibana_context.savedSearchId');
+      if (reference) {
+        state.savedSearchId[0] = reference.id;
+      }
+      return state;
     },
-  },
 
-  async fn(input, args, { getSavedObject }) {
-    const timeRange = args.timeRange || input?.timeRange;
-    let queries = mergeQueries(input?.query, args?.q || []);
-    let filters = [...(input?.filters || []), ...(args?.filters?.map(unboxExpressionValue) || [])];
+    async fn(input, args, { getKibanaRequest }) {
```

Comment:
I'm a bit confused about what does it resolve to, as `KibanaRequest` is the server-side type and makes sense only on the server.

Not specific to this PR:

Also, I'm skeptical about using `KibanaRequest` type in `expressions` plugin. Mayb there is a way to create our own wrapper around it, call it `ExpressionUserContext`.

The problem I see with it is that it assumes that you can authenticate only using `KibanaRequest`. Also, it exports `KibanaRequest` in `ExecutionContext` interface from `expressions` plugin, making it our public API, but when Core will want to change something in `KibanaRequest`, they might need to do changes to `expressions` plugin and it might be breaking change in `expressions` plugin. Just some thoughts.

Reply:
getKibanaRequest on the client will be undefined, but the getStartServices function on the client doesn't expect it. this is just so we can keep the same code in common.

i agree the whole kibana request thing is a bit weird and we should probably do somethin about it in the future